atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.processors;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.PackageElement;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+
+import com.google.common.collect.Sets;
+
+final class TypeToExtend {
+    private TypeElement typeToExtend;
+    private PackageElement typePackage;
+    private Set<ExecutableElement> methods;
+    private Set<ExecutableElement> constructors;
+
+    TypeToExtend(PackageElement typePackage,
+            TypeElement typeToExtend,
+            TypeElement... supertypes) {
+
+        this.typeToExtend = typeToExtend;
+        this.typePackage = typePackage;
+
+        List<ExecutableElement> allMethods = extractMethods(typeToExtend);
+        for (TypeElement supertype : supertypes) {
+            allMethods.addAll(extractMethods(supertype));
+        }
+
+        Map<String, ExecutableElement> unifiedMethods = allMethods
+                .stream()
+                .collect(Collectors.toMap(ExecutableElement::toString, Function.identity(), (name1, name2) -> name1));
+
+        methods = Sets.newHashSet(unifiedMethods.values());
+        constructors = extractConstructors(typeToExtend);
+    }
+
+    private List<ExecutableElement> extractMethods(TypeElement typeToExtractMethodsFrom) {
+        Predicate<Element> filterForType;
+        if (typeToExtractMethodsFrom.getKind() == ElementKind.INTERFACE) {
+            filterForType = this::interfaceMethodFilter;
+        } else {
+            filterForType = this::classMethodFilter;
+        }
+
+        return typeToExtractMethodsFrom.getEnclosedElements()
+                .stream()
+                .filter(filterForType)
+                .map(element -> (ExecutableElement) element)
+                .collect(Collectors.toList());
+    }
+
+    private Boolean interfaceMethodFilter(Element element) {
+        return element.getKind() == ElementKind.METHOD
+                && element.getModifiers().contains(Modifier.PUBLIC);
+    }
+
+    private Boolean classMethodFilter(Element element) {
+        return element.getKind() == ElementKind.METHOD
+                && element.getModifiers().contains(Modifier.PUBLIC)
+                && !element.getModifiers().contains(Modifier.FINAL)
+                && !element.getModifiers().contains(Modifier.NATIVE)
+                && !element.getModifiers().contains(Modifier.STATIC)
+                && !element.getSimpleName().contentEquals("equals")
+                && !element.getSimpleName().contentEquals("toString");
```

Comment:
`Object.java`'s `hashCode` is native, so it's filtered by the NATIVE check above. But agreed that hashCode could also be filtered with `equals` and `toString`. Added.